File: apps/desktop/src-tauri/src/items.rs
//! What the game says about an item the addon only recorded the number of.
//!
//! A segment arrives carrying item ids — the transmog sources a character learned, the pieces
//! an equipment set was changed to hold — and, where the client happened to have the item
//! loaded at the moment, a name. Everything a player recognises an item by is in two of the
//! client's own tables: `Item` is what kind of thing it is, where it is worn and the picture
//! beside it, and `ItemSparse` is what it is called, what it is worth and who may wear it.
//! [`read`] joins the two for the handful of ids a window is showing.
//!
//! Nothing here decides wording. Which subclass of armour is "Leather" and which slot is
//! "Head" is the window's business, and doing it there keeps the language beside the markup
//! it goes in; this answers with the numbers the game stores and no more.
//!
//! The column positions are [`crate::tables`]'s, out of `docs/game-tables.json`, and were read
//! off build 12.0.5.67 with `examples/dump_item_facts` — which finds them rather than trusting
//! them and is what to run again after a patch. A patch that reorders one shows wrong values
//! rather than failing, so every position the registry holds for these two tables is one that
//! tool checks against something a wrong column could not produce.

use std::collections::HashMap;
use std::sync::Mutex;

use serde::Serialize;
use serde_json::{json, Value};

use crate::casc::GameFiles;
use crate::db2::Db2;
use crate::tables::item as column;
use crate::tables::item_sparse as item_column;
use crate::tables::{ITEM, ITEM_SPARSE};

/// What the game files armour and weapons under, which is all this module names by hand.
pub const ARMOR: u32 = 4;
pub const WEAPON: u32 = 2;

/// The class mask on an item anybody may wear, which is nearly all of them.
///
/// Stored as a signed 16-bit `-1` and read back out of a bit field, so it arrives as this
/// rather than as something with every bit set.
pub const ANY_CLASS: u32 = 0xFFFF;

/// One item, in the facts the game keeps about it.
///
/// Everything here is a number the game stores except the name, and the window turns them
/// into words. An item this install cannot describe has no `Item` at all rather than an
/// `Item` full of zeroes — see [`read`].
#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct Item {
    pub id: u32,
    /// What the game calls it. Empty for an item `ItemSparse` holds no readable row for — or
    /// holds one that names it nothing, which comes to the same thing — and that is a real
    /// answer: what `Item` said is still worth drawing, and the window falls back to the name
    /// the addon caught. Nothing else from an unnamed row is carried; see [`read`].
    pub name: String,
    /// [`ARMOR`], [`WEAPON`], or one of the kinds nothing is worn from.
    pub class_id: u32,
    pub subclass_id: u32,
    /// Where it is worn: 1 head, 5 chest, 13 a one-hander, 17 a two-hander. Zero for a thing
    /// that is not worn at all.
    pub inventory_type: u32,
    /// The colour the game writes it in: 0 poor, 1 common, 2 uncommon, 3 rare, 4 epic,
    /// 5 legendary, 6 artifact, 7 heirloom.
    pub quality: u32,
    /// The level needed to equip it. Zero is the ordinary answer.
    pub required_level: u32,
    /// A bit per class, in the game's class order, or [`ANY_CLASS`] for anybody.
    pub allowable_class: u32,
    /// The picture beside it, as a FileDataID to be decoded through `icons`. Zero where the
    /// row names none.
    pub icon_file_data_id: u32,
}

/// Looks up the items a window is showing, and nothing else.
///
/// Two tables, opened in the order that costs least: `Item` is two megabytes and answers for
/// every item the install holds, and `ItemSparse` is sixty-three and is opened only once
/// something has been found to name. A batch nothing at all was found for never opens it.
///
/// `None` against an id is an answer rather than a failure, the way it is for an achievement:
/// the game encrypts content it has not shipped, so an item from a build newer than this
/// install — or one a patch removed — is a row that is simply not there, and the rest of a
/// segment's list still has to draw.
pub fn read(files: &dyn GameFiles, wanted: &[u32]) -> Result<Vec<(u32, Option<Item>)>, String> {
    if wanted.is_empty() {
        return Ok(Vec::new());
    }

    let table = Db2::parse(files.read(ITEM)?)?;
    let mut found: HashMap<u32, Item> = table
        .rows()
        .filter(|row| wanted.contains(&row.id()))
        .map(|row| {
            let id = row.id();
            (
                id,
                Item {
                    id,
                    name: String::new(),
                    class_id: row.number(column::CLASS),
                    subclass_id: row.number(column::SUBCLASS),
                    inventory_type: row.number(column::INVENTORY_TYPE),
                    quality: 0,
                    required_level: 0,
                    allowable_class: ANY_CLASS,
                    icon_file_data_id: row.number(column::ICON_FILE_ID),
                },
            )
        })
        .collect();

    if !found.is_empty() {
        let sparse = Db2::parse_with_text_columns(files.read(ITEM_SPARSE)?, &item_column::TEXT)?;
        for row in sparse.rows() {
            let Some(item) = found.get_mut(&row.id()) else {
                continue;
            };
            // A row that cannot say what the item is called says nothing else either. The name
            // and the numbers are the same record read at different offsets, so whatever leaves
            // the text empty — a patch that moved the strings, a locale that holds none — leaves
            // the numbers pointing somewhere they should not, and a quality read as a required
            // level is worse than no level at all: it draws as a fact. The window makes the same
            // judgement about the same row when it declines to colour an item it cannot name.
            let name = row.text(item_column::NAME);
            if name.is_empty() {
                continue;
            }
            item.name = name;
            item.quality = row.number(item_column::QUALITY);
            item.required_level = row.number(item_column::REQUIRED_LEVEL);
            item.allowable_class = row.number(item_column::ALLOWABLE_CLASS);
        }
    }

    Ok(wanted
        .iter()
        .map(|id| (*id, found.get(id).cloned()))
        .collect())
}

/// The items looked up so far, kept for as long as the app runs.
///
/// Same bargain as the achievement book, and a better one: what the game says about an item
/// cannot change under a running app, a reader walking their history meets the same items
/// over and over, and the read behind this one opens sixty-three megabytes of `ItemSparse`.
/// An id the tables answered nothing for is remembered as such, because that too is a fact
/// about the install rather than about the moment.
#[derive(Default)]
pub struct ItemBook {
    known: Mutex<HashMap<u32, Option<Item>>>,
}

impl ItemBook {
    /// Which of the ids asked for have not been looked up yet, without repeats.
    pub fn missing(&self, wanted: &[u32]) -> Vec<u32> {
        let known = self.known.lock().expect("the item book is not poisoned");
        let mut missing: Vec<u32> = Vec::new();
        for id in wanted {
            if *id != 0 && !known.contains_key(id) && !missing.contains(id) {
                missing.push(*id);
            }
        }
        missing
    }

    pub fn store(&self, found: Vec<(u32, Option<Item>)>) {
        let mut known = self.known.lock().expect("the item book is not poisoned");
        known.extend(found);
    }

    /// The answer to a request: the items among those asked for that this install can
    /// describe, keyed by the id the segment named them by.
    ///
    /// The ones it cannot are left out rather than sent as null, so a row draws exactly as it
    /// would have before the lookup came back — the addon's own name, and no more.
    pub fn answer(&self, wanted: &[u32]) -> Value {
        let known = self.known.lock().expect("the item book is not poisoned");
        let mut items = serde_json::Map::new();
        for id in wanted {
            if let Some(Some(found)) = known.get(id) {
                items.insert(id.to_string(), json!(found));
            }
        }
        json!({ "items": Value::Object(items) })
    }
}

#[cfg(test)]
mod tests {
    use std::cell::RefCell;

    use super::*;
    use crate::casc::{item_fixture_files, DirFiles};

    /// The invented items, by the id a segment would name them under.
    /// A pair of leather shoulders, which is the ordinary case.
    const WANDERERS_MANTLE: u32 = 201;
    /// Plate, and the only item restricted to some classes.
    const BULWARK_HELM: u32 = 202;
    /// A one-handed sword, so a weapon's slot has to come from the item rather than the
    /// appearance.
    const TIDEGLASS_EDGE: u32 = 203;
    /// A cloak: cloth by class, worn on the back.
    const NIGHT_CLOAK: u32 = 204;
    /// Held in `Item` and not in `ItemSparse`, which is half a row rather than none.
    const NAMELESS_TRINKET: u32 = 205;
    /// Held in both, and named by neither: an `ItemSparse` row whose text is empty and whose
    /// numbers are not.
    const UNNAMED_CHEST: u32 = 207;
    /// Not worn at all, so nothing about it is a slot.
    const HEARTH_TOKEN: u32 = 206;
    /// A mail chestpiece the game keeps encrypted.
    const WITHHELD: u32 = 900;
    /// An id no row of either table carries.
    const ABSENT: u32 = 4242;

    /// Game files that remember what was asked of them. Which table a read opens is part of
    /// the behaviour: the big one is only opened when there is something to name.
    struct Noted {
        files: DirFiles,
        asked: RefCell<Vec<u32>>,
    }

    impl Noted {
        fn new() -> Self {
            Self {
                files: item_fixture_files(),
                asked: RefCell::new(Vec::new()),
            }
        }
    }

    impl GameFiles for Noted {
        fn read(&self, fdid: u32) -> Result<std::sync::Arc<Vec<u8>>, String> {
            self.asked.borrow_mut().push(fdid);
            self.files.read(fdid)
        }
    }

    fn read_all(wanted: &[u32]) -> Vec<(u32, Option<Item>)> {
        read(&item_fixture_files(), wanted).unwrap()
    }

    fn one(id: u32) -> Item {
        read_all(&[id])
            .into_iter()
            .next()
            .and_then(|(_, found)| found)
            .unwrap_or_else(|| panic!("the fixture holds item {id}"))
    }

    // Everything the window shows about an item comes out of one row of each table, so the
    // whole of it is written out here rather than a field at a time.
    #[test]
    fn reads_an_item_down_to_what_the_game_calls_it() {
        assert_eq!(
            one(WANDERERS_MANTLE),
            Item {
                id: 201,
                name: "Wanderer's Mantle".into(),
                class_id: ARMOR,
                subclass_id: 2,
                inventory_type: 3,
                quality: 3,
                required_level: 25,
                allowable_class: ANY_CLASS,
                icon_file_data_id: 260001,
            }
        );
    }

    // The two tables are joined on the item id, and a reader that took either half positionally
    // would put one item's name on another's row.
    #[test]
    fn keeps_each_items_name_on_its_own_row() {
        let named: Vec<(u32, String)> = read_all(&[TIDEGLASS_EDGE, NIGHT_CLOAK, BULWARK_HELM])
            .into_iter()
            .filter_map(|(id, found)| found.map(|found| (id, found.name)))
            .collect();
        assert_eq!(
            named,
            vec![
                (TIDEGLASS_EDGE, "Tideglass Edge".to_string()),
                (NIGHT_CLOAK, "Cloak of the Long Night".to_string()),
                (BULWARK_HELM, "Bulwark Helm".to_string()),
            ]
        );
    }

    // Which armour class a piece is, and which hand a weapon goes in, is the pair of facts the
    // whole lookup exists for: the addon records neither.
    #[test]
    fn says_what_kind_of_thing_an_item_is_and_where_it_is_worn() {
        let kinds: Vec<(u32, u32, u32, u32)> =
            read_all(&[WANDERERS_MANTLE, BULWARK_HELM, TIDEGLASS_EDGE, HEARTH_TOKEN])
                .into_iter()
                .filter_map(|(id, found)| {
                    found.map(|found| (id, found.class_id, found.subclass_id, found.inventory_type))
                })
                .collect();
        assert_eq!(
            kinds,
            vec![
                (WANDERERS_MANTLE, ARMOR, 2, 3), // leather shoulders
                (BULWARK_HELM, ARMOR, 4, 1),     // plate head
                (TIDEGLASS_EDGE, WEAPON, 7, 13), // a one-handed sword
                (HEARTH_TOKEN, 15, 0, 0),        // worn nowhere at all
            ]
        );
    }

    // Nearly every item is for every class, so the game keeps the exception sparsely: a reader
    // that missed it would report the whole game as being restricted to whichever class is
    // numbered first.
    #[test]
    fn says_which_classes_may_wear_an_item_when_it_is_not_all_of_them() {
        // Warrior, paladin and death knight — the three that wear plate — as a bit each.
        assert_eq!(one(BULWARK_HELM).allowable_class, 0b10_0011);
        assert_eq!(one(WANDERERS_MANTLE).allowable_class, ANY_CLASS);
        assert_eq!(one(TIDEGLASS_EDGE).allowable_class, ANY_CLASS);
    }

    // The colour an item's name is written in and the level it takes to wear it both come out
    // of the big table, and both are worth nothing more than the number the game stores.
    #[test]
    fn reads_what_an_item_is_worth_and_what_it_takes_to_wear_it() {
        let worth: Vec<(u32, u32, u32)> =
            read_all(&[WANDERERS_MANTLE, BULWARK_HELM, TIDEGLASS_EDGE, HEARTH_TOKEN])
                .into_iter()
                .filter_map(|(id, found)| {
                    found.map(|found| (id, found.quality, found.required_level))
                })
                .collect();
        assert_eq!(
            worth,
            vec![
                (WANDERERS_MANTLE, 3, 25),
                (BULWARK_HELM, 4, 60),
                (TIDEGLASS_EDGE, 5, 60),
                // A token nobody wears: common, and needing no level at all.
                (HEARTH_TOKEN, 1, 0),
            ]
        );
    }

    // A row in the small table and none in the big one is half an answer, and half an answer
    // is worth drawing: the slot and the picture are there, and the window falls back to the
    // name the addon caught.
    #[test]
    fn answers_with_what_it_has_for_an_item_the_big_table_cannot_name() {
        let found = one(NAMELESS_TRINKET);
        assert_eq!(found.name, "");
        assert_eq!(found.inventory_type, 12);
        assert_eq!(found.icon_file_data_id, 260004);
        assert_eq!(found.quality, 0);
    }

    // A row that cannot say what the item is called is not a row to read numbers off either.
    //
    // The window showed "Item 39270 · Sword · One-hand · Level 4" about an epic Wrath sword that
    // takes level 30: the name had come back empty and the level was the quality, read a few
    // columns off. Whatever moves the strings moves the numbers beside them, and a level that is
    // really a quality is worse than no level — it reads as a fact. So an unnamed row contributes
    // nothing but its silence, and what the row *before* it said still stands: the slot, the
    // kind and the picture come out of `Item` and are unaffected.
    //
    // `itemLine` in the window already refuses to colour an item it cannot name, for the same
    // reason and about the same row.
    #[test]
    fn takes_no_numbers_from_a_row_that_cannot_name_the_item() {
        let found = one(UNNAMED_CHEST);
        assert_eq!(found.name, "");
        // What `Item` says, which is not in doubt.
        assert_eq!(
            (found.class_id, found.subclass_id, found.inventory_type),
            (ARMOR, 1, 5)
        );
        assert_eq!(found.icon_file_data_id, 260003);
        // And what the nameless row said, which is not believed: the defaults an item nothing is
        // known about carries.
        assert_eq!(found.quality, 0);
        assert_eq!(found.required_level, 0);
        assert_eq!(found.allowable_class, ANY_CLASS);
    }

    // Two ways an id can go unanswered, and neither is a reason to fail the batch.
    #[test]
    fn answers_with_nothing_for_an_item_this_install_cannot_describe() {
        let named: Vec<(u32, bool)> = read_all(&[WANDERERS_MANTLE, WITHHELD, ABSENT])
            .iter()
            .map(|(id, found)| (*id, found.is_some()))
            .collect();
        assert_eq!(
            named,
            vec![(WANDERERS_MANTLE, true), (WITHHELD, false), (ABSENT, false)]
        );
    }

    // Both tables are parsed once however many items a batch asks about, and the answer keeps
    // the order it was asked in.
    #[test]
    fn opens_each_table_once_for_a_whole_batch() {
        let files = Noted::new();
        let found = read(&files, &[NIGHT_CLOAK, WANDERERS_MANTLE, TIDEGLASS_EDGE]).unwrap();
        assert_eq!(
            found.iter().map(|(id, _)| *id).collect::<Vec<u32>>(),
            vec![NIGHT_CLOAK, WANDERERS_MANTLE, TIDEGLASS_EDGE]
        );
        assert_eq!(files.asked.into_inner(), vec![ITEM, ITEM_SPARSE]);
    }

    // The big table is sixty-three megabytes on a real install, and a batch with nothing to
    // name has nothing to look up in it.
    #[test]
    fn does_not_open_the_big_table_for_a_batch_it_can_describe_nothing_of() {
        let files = Noted::new();
        assert_eq!(read(&files, &[WITHHELD, ABSENT]).unwrap().len(), 2);
        assert_eq!(files.asked.into_inner(), vec![ITEM]);
    }

    #[test]
    fn reads_nothing_at_all_when_nothing_was_asked_for() {
        let files = Noted::new();
        assert_eq!(read(&files, &[]).unwrap(), Vec::new());
        assert_eq!(files.asked.into_inner(), Vec::<u32>::new());
    }

    #[test]
    fn says_so_when_a_table_is_not_there() {
        let temp = tempfile::tempdir().unwrap();
        let error = read(&DirFiles::new(temp.path()), &[WANDERERS_MANTLE]).unwrap_err();
        assert!(error.contains("841626.db2"), "{error}");
    }

    /* ---------- the book ---------- */

    // A history mentions the same item once per segment it turned up in, and the read behind
    // this opens sixty-three megabytes — so an id is looked up once and not again.
    #[test]
    fn looks_an_item_up_the_first_time_it_is_named_and_not_again() {
        let files = Noted::new();
        let book = ItemBook::default();

        let first = book.missing(&[WANDERERS_MANTLE, BULWARK_HELM]);
        assert_eq!(first, vec![WANDERERS_MANTLE, BULWARK_HELM]);
        book.store(read(&files, &first).unwrap());

        let second = book.missing(&[WANDERERS_MANTLE, TIDEGLASS_EDGE]);
        assert_eq!(second, vec![TIDEGLASS_EDGE]);
        book.store(read(&files, &second).unwrap());

        assert_eq!(
            files.asked.into_inner(),
            vec![ITEM, ITEM_SPARSE, ITEM, ITEM_SPARSE]
        );
        let answer = book.answer(&[WANDERERS_MANTLE, BULWARK_HELM, TIDEGLASS_EDGE]);
        assert_eq!(answer["items"].as_object().unwrap().len(), 3);
    }

    // Whether an install can describe an item is a fact about the install, so asking again
    // would cost the same two reads to arrive back at the same nothing.
    #[test]
    fn does_not_go_looking_again_for_an_item_it_already_failed_to_find() {
        let book = ItemBook::default();
        book.store(read(&item_fixture_files(), &book.missing(&[WITHHELD, ABSENT])).unwrap());

        assert_eq!(book.missing(&[WITHHELD, ABSENT]), Vec::<u32>::new());
        assert_eq!(book.answer(&[WITHHELD, ABSENT])["items"], json!({}));
    }

    // A window asks with the list its segments carry, which repeats an item collected on two
    // characters and can hold the zero an event with no id at all comes across as.
    #[test]
    fn asks_after_one_item_however_many_segments_name_it() {
        let book = ItemBook::default();
        assert_eq!(
            book.missing(&[BULWARK_HELM, BULWARK_HELM, 0, NIGHT_CLOAK, BULWARK_HELM]),
            vec![BULWARK_HELM, NIGHT_CLOAK]
        );
    }

    // The window keys what it draws by the id the segment carries, so the answer is keyed the
    // same way rather than positionally — half of a request can be missing.
    #[test]
    fn keys_what_it_answers_with_by_the_id_the_segment_named() {
        let book = ItemBook::default();
        book.store(read(&item_fixture_files(), &[WANDERERS_MANTLE, WITHHELD]).unwrap());
        let answer = book.answer(&[WANDERERS_MANTLE, WITHHELD]);
        assert_eq!(answer["items"]["201"]["name"], json!("Wanderer's Mantle"));
        assert_eq!(answer["items"]["201"]["subclassId"], json!(2));
        assert_eq!(answer["items"]["201"]["inventoryType"], json!(3));
        assert_eq!(answer["items"]["201"]["iconFileDataId"], json!(260001));
        assert_eq!(answer["items"][WITHHELD.to_string()], Value::Null);
    }
}

File: apps/desktop/src/items.ts
/**
 * What the game says about an item, kept beside what the addon recorded of it.
 *
 * A segment carries item ids — the transmog sources a character learned, the pieces an
 * equipment set changed to hold — and, when the client happened to have the item loaded at the
 * moment, a name. Everything a player recognises an item by is in the game's own tables: what
 * it is called, what it is worth, what kind of armour it is, where it is worn, who may wear it
 * and the picture beside it. The backend reads those and answers with numbers; this is where
 * the numbers become the words on screen.
 *
 * [`itemLine`] is what a row makes of the two halves, including when the second half never
 * arrives — an install can say nothing about an item it does not have, and a row still has to
 * draw. [`createItemBook`] is what fetches and remembers them.
 */

import type { IconsPayload, ItemDetail, ItemDetailsPayload } from "./types";

/** The class of thing an item is, in the game's own numbering. */
const WEAPON = 2;
const ARMOR = 4;

/** A class mask nobody is excluded by, which is what nearly every item carries. */
const ANY_CLASS = 0xffff;

/**
 * The armour classes, by the subclass the game files a piece under.
 *
 * Subclass 0 is deliberately absent. It is what rings, necks and trinkets are filed as —
 * "Miscellaneous", which is the game admitting the question does not apply — and a chip
 * saying so on every ring would be noise rather than information.
 */
const ARMOR_CLASSES: Record<number, string> = {
  1: "Cloth",
  2: "Leather",
  3: "Mail",
  4: "Plate",
  5: "Cosmetic",
  6: "Shield",
};

/**
 * What kind of weapon it is.
 *
 * The game keeps one-handed and two-handed axes, maces and swords as separate subclasses;
 * they are collapsed here because the slot beside them already says which, and "Two-hand ·
 * Sword (2H)" says it twice.
 */
const WEAPON_KINDS: Record<number, string> = {
  0: "Axe",
  1: "Axe",
  2: "Bow",
  3: "Gun",
  4: "Mace",
  5: "Mace",
  6: "Polearm",
  7: "Sword",
  8: "Sword",
  9: "Warglaive",
  10: "Staff",
  13: "Fist weapon",
  15: "Dagger",
  16: "Thrown",
  18: "Crossbow",
  19: "Wand",
  20: "Fishing pole",
};

/**
 * Where the game says an item is worn.
 *
 * 0 is not in here and neither are the kinds nothing is worn from: a hearthstone, a bag of
 * flour and a quest item are all "worn nowhere", which is a fact about the item rather than a
 * slot to name.
 */
const SLOTS: Record<number, string> = {
  1: "Head",
  2: "Neck",
  3: "Shoulders",
  4: "Shirt",
  5: "Chest",
  6: "Waist",
  7: "Legs",
  8: "Feet",
  9: "Wrists",
  10: "Hands",
  11: "Ring",
  12: "Trinket",
  13: "One-hand",
  14: "Off hand",
  15: "Ranged",
  16: "Back",
  17: "Two-hand",
  19: "Tabard",
  20: "Chest",
  21: "Main hand",
  22: "Off hand",
  23: "Held in off hand",
  26: "Ranged",
  28: "Relic",
  29: "Profession tool",
  30: "Profession accessory",
};

/** What the game is worth calling each quality, for the ones worth saying out loud. */
const QUALITIES: Record<number, string> = {
  0: "Poor",
  1: "Common",
  2: "Uncommon",
  3: "Rare",
  4: "Epic",
  5: "Legendary",
  6: "Artifact",
  7: "Heirloom",
};

/**
 * The classes, in the order the game numbers them: the mask holds `1 << (class - 1)`.
 *
 * Every class the game has shipped is here, and the mask on a legacy item can still carry
 * bits above them — the ones set aside for classes that never arrived. Those are ignored
 * rather than named, which is why a restriction is built from this list rather than from the
 * bits that happen to be set.
 */
const CLASSES = [
  "Warrior", "Paladin", "Hunter", "Rogue", "Priest", "Death Knight", "Shaman", "Mage",
  "Warlock", "Monk", "Druid", "Demon Hunter", "Evoker",
];

/** One item as a row reads it, with everything the markup needs already decided. */
export interface ItemLine {
  /** What names it: the game's own name, else the name the addon caught, else the id. */
  name: string;
  /**
   * The colour the game writes the name in, or null for an item nothing is known about.
   *
   * Null rather than zero, because zero is poor quality — the grey of a broken sword — and an
   * item whose lookup has not come back is not grey, it is uncoloured.
   */
  quality: number | null;
  /** "Epic", "Rare". Empty until the game has been asked. */
  qualityName: string;
  /** What kind of thing it is: "Leather", "Plate", "Sword". Empty when the game files it
   * under nothing worth saying. */
  kind: string;
  /** Where it is worn: "Shoulders", "Two-hand". Empty for a thing that is not worn. */
  slot: string;
  /** Who may wear it, when it is not everybody: "Warrior, Paladin, Death Knight only". */
  restriction: string;
  /** The level it takes, when it takes one: "Level 60". */
  requirement: string;
  /** The picture the game shows beside it, or zero when there is none to ask for. */
  iconFileDataId: number;
}

/**
 * What an item is called, wherever in the app it is named.
 *
 * The game's name leads and the addon's is the fallback, the same way round as an
 * achievement's: the game's is the one that is definitely spelled the way the game spells it,
 * while the addon's was whatever the client had loaded at the time — and for an item the
 * install cannot describe at all, the addon's name is the only name there is. Under both of
 * them is the id, which is at least something a reader can look up.
 *
 * One function rather than the four copies of `name || "Item " + id` this app grew, because
 * the copies disagreed: a summary chip worded before anything had been asked said "Item 39473"
 * while the list it unfolded into said "Insanity's Grip", about the same piece, on the same
 * card. Anything that has to say what an item is called says it through here.
 */
export const itemName = (id: number, recorded?: string | null, detail?: ItemDetail): string =>
  detail?.name || recorded || `Item ${id}`;

/**
 * How a row reads, out of what the addon recorded and what the game could be asked.
 */
export function itemLine(
  id: number,
  recorded?: string | null,
  detail?: ItemDetail,
): ItemLine {
  const quality = detail && detail.name ? detail.quality : null;
  return {
    name: itemName(id, recorded, detail),
    quality,
    qualityName: quality == null ? "" : QUALITIES[quality] ?? "",
    kind: kindOf(detail),
    slot: detail ? SLOTS[detail.inventoryType] ?? "" : "",
    restriction: restrictionOf(detail),
    requirement: detail && detail.requiredLevel > 0 ? `Level ${detail.requiredLevel}` : "",
    iconFileDataId: detail?.iconFileDataId ?? 0,
  };
}

/** Which armour class or which weapon, and nothing at all for the kinds that are neither. */
function kindOf(detail?: ItemDetail): string {
  if (!detail) return "";
  if (detail.classId === ARMOR) return ARMOR_CLASSES[detail.subclassId] ?? "";
  if (detail.classId === WEAPON) return WEAPON_KINDS[detail.subclassId] ?? "";
  return "";
}

/**
 * Who may wear it, named class by class, and empty when anybody may.
 *
 * The mask is read against the classes the game has rather than bit by bit: a legacy item can
 * have bits set for classes that never shipped, and a mask covering every class that exists
 * is not a restriction however many spare bits it carries.
 */
function restrictionOf(detail?: ItemDetail): string {
  if (!detail) return "";
  const mask = detail.allowableClass;
  if (mask === 0 || mask === ANY_CLASS) return "";
  const named = CLASSES.filter((_, index) => (mask & (1 << index)) !== 0);
  if (!named.length || named.length === CLASSES.length) return "";
  return `${named.join(", ")} only`;
}

export interface ItemBookOptions {
  /** Asks the backend what the game says about a list of ids. */
  load: (ids: number[]) => Promise<ItemDetailsPayload>;
  /** Asks the backend for the pictures those items name. */
  loadIcons: (iconFileDataIds: number[]) => Promise<IconsPayload>;
  /**
   * Runs the batch, once everything asking in this turn has asked. A microtask by default,
   * and a hook a test can hold open to see what one request ended up carrying.
   */
  schedule?: (run: () => void) => void;
}

export interface ItemBook {
  /**
   * Puts `ids` in the next request and watches for what comes back, until the function it
   * hands back is called.
   *
   * Every item on screen calls this for itself, which is why the request is batched rather
   * than sent: twenty items in a segment are one lookup, and the read behind it opens the
   * game's sixty-three megabyte table once for however many ids it carries. `changed` fires
   * when the words arrive and again when the pictures do — two reads, and a list of items is
   * worth reading while the second is still going — and it fires for a row whose item some
   * other row asked after first, which is what makes the same item on two rows draw twice
   * rather than once.
   *
   * The unsubscribe shape is React's own: an effect returns it and the row stops listening
   * when it leaves the screen.
   */
  learn: (ids: number[], changed: () => void) => () => void;
  detail: (id: number) => ItemDetail | undefined;
  /** The picture for an item, once it has arrived. */
  icon: (id: number) => string | undefined;
}

/**
 * The items this window has been told about, and their pictures.
 *
 * What the game says about an item cannot change under a running app, and a reader walking
 * their history meets the same items over and over — a set collected piece by piece, a
 * favourite weapon transmogged onto everything — so an id is asked about once. The backend
 * remembers them too; this saves the round trip as well.
 *
 * A lookup that fails is forgotten rather than remembered as nothing: the reasons are the ones
 * that stop the whole game folder being readable — it has not been chosen yet, or it is
 * mid-patch — and those are worth one more try when the reader opens the next segment. It is
 * never reported, because a row that says what the addon recorded is what the app showed
 * before any of this, and an apology in its place would be worse.
 */
export function createItemBook(
  { load, loadIcons, schedule = queueMicrotask }: ItemBookOptions,
): ItemBook {
  const known = new Map<number, ItemDetail>();
  const icons = new Map<number, string>();
  /** Ids a request has already been made for, whatever it came back with. */
  const asked = new Set<number>();
  /** Textures a request has already been made for, likewise. */
  const askedIcons = new Set<number>();

  /** What the next request will carry, and everything on screen waiting to hear about it. */
  let pending = new Set<number>();
  let sending = false;
  const listeners = new Set<() => void>();

  /**
   * Says that something new has arrived, to everything currently on screen.
   *
   * To all of them rather than to whoever asked, because who asked is not who is waiting: the
   * same item can be on two rows, and only the first of them puts it in the request.
   */
  const tell = (): void => {
    for (const listener of [...listeners]) listener();
  };

  /** Sends whatever has piled up since the last request, as one request. */
  async function send(ids: number[]): Promise<void> {
    try {
      const payload = await load(ids);
      for (const [id, detail] of Object.entries(payload.items ?? {})) known.set(Number(id), detail);
    } catch {
      for (const id of ids) asked.delete(id);
      return;
    }
    tell();

    const pictures = [...new Set(ids.map((id) => known.get(id)?.iconFileDataId ?? 0))]
      .filter((fdid) => fdid > 0 && !askedIcons.has(fdid));
    if (!pictures.length) return;
    for (const fdid of pictures) askedIcons.add(fdid);
    try {
      const payload = await loadIcons(pictures);
      for (const [fdid, url] of Object.entries(payload.icons ?? {})) icons.set(Number(fdid), url);
    } catch {
      for (const fdid of pictures) askedIcons.delete(fdid);
      return;
    }
    tell();
  }

  function learn(ids: number[], changed: () => void): () => void {
    listeners.add(changed);
    for (const id of ids) {
      if (id > 0 && !asked.has(id)) {
        asked.add(id);
        pending.add(id);
      }
    }
    if (pending.size && !sending) {
      sending = true;
      schedule(() => {
        const carrying = [...pending];
        pending = new Set();
        sending = false;
        void send(carrying);
      });
    }
    return () => listeners.delete(changed);
  }

  return {
    learn,
    detail: (id) => known.get(id),
    icon: (id) => {
      const fdid = known.get(id)?.iconFileDataId;
      return fdid ? icons.get(fdid) : undefined;
    },
  };
}
